pkg/kudoctl/packages/template/template_parsing.go

```diff
@@ -0,0 +1,135 @@
+package template
+
+import (
+	"fmt"
+	"strings"
+	"text/template/parse"
+
+	"github.com/kudobuilder/kudo/pkg/engine/renderer"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/clog"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/packages"
+)
+
+// nodes are template nodes
+type nodes struct {
+	parameters     []string
+	implicitParams []string
+	error          *string
+}
+
+// nodeMap is a map of template files to template nodes
+type nodeMap map[string]nodes
+
+// getNodeMap converts a set of Templates to nodeMap which is a map of file names to template nodes
+func getNodeMap(ts packages.Templates) nodeMap {
+
+	tNodes := nodeMap{}
+
+	for fname, file := range ts {
+
+		e := renderer.New()
+		t := e.Template(fname)
+		// parse 1 template file using engine render template
+		tplate, err := t.Parse(file)
+		if err != nil {
+			errMsg := fmt.Sprintf("template file %q reports the following error: %v", fname, err)
+			n := nodes{
+				error: &errMsg,
+			}
+			tNodes[fname] = n
+
+			continue
+		}
+
+		//nodeMap is a map of node types ("Implicits", "Params") to a set of that type (which is go is a map :))
+		nodeMap := map[string]map[string]bool{}
+		walkNodes(tplate.Root, fname, nodeMap)
+
+		n := nodes{
+			parameters:     values(nodeMap, "Params"),
+			implicitParams: values(nodeMap, "Implicits"),
+		}
+		tNodes[fname] = n
+	}
+
+	return tNodes
+}
+
+//walkNodes walks the nodes of a template providing an array of parameters
+func walkNodes(node parse.Node, fname string, nodeMap map[string]map[string]bool) {
```

Comment:
this is very go-template internals thing, I don't really understand all the implications of how are certain types parsed. Maybe this deserves more comments/docs or some references to go template docs?